apps/api/R/inputs.R

```diff
@@ -0,0 +1,172 @@
+    inputs <- inputs %>%
+      filter(site_id == !!site_id)
+  }
+  
+  if(! is.null(format_id)) {
+    inputs <- inputs %>%
+      filter(format_id == !!format_id)
+  }
+  
+  if(! is.null(host_id)) {
+    inputs <- inputs %>%
+      filter(machine_id == !!host_id)
+  }
+  
+  qry_res <- inputs %>%
+    select(-site_id, -model_id, -format_id, -machine_id) %>%
+    distinct() %>%
+    arrange(id) %>%
+    collect()
+  
+  PEcAn.DB::db.close(dbcon)
+  
+  if (nrow(qry_res) == 0 || as.numeric(offset) >= nrow(qry_res)) {
+    res$status <- 404
+    return(list(error="Input(s) not found"))
+  }
+  else {
+    has_next <- FALSE
+    has_prev <- FALSE
+    if (nrow(qry_res) > (as.numeric(offset) + as.numeric(limit))) {
+      has_next <- TRUE
+    }
+    if (as.numeric(offset) != 0) {
+      has_prev <- TRUE
+    }
+    
+    qry_res <- qry_res[(as.numeric(offset) + 1):min((as.numeric(offset) + as.numeric(limit)), nrow(qry_res)), ]
+    
+    result <- list(inputs = qry_res)
+    result$count <- nrow(qry_res)
+    if(has_next){
+      result$next_page <- paste0(
+        req$rook.url_scheme, "://",
+        req$HTTP_HOST,
+        "/api/workflows",
+        req$PATH_INFO,
+        substr(req$QUERY_STRING, 0, stringr::str_locate(req$QUERY_STRING, "offset=")[[2]]),
+        (as.numeric(limit) + as.numeric(offset)),
+        "&limit=", 
+        limit
+      )
+    }
+    if(has_prev) {
+      result$prev_page <- paste0(
+        req$rook.url_scheme, "://",
+        req$HTTP_HOST,
+        "/api/workflows",
+        req$PATH_INFO, 
+        substr(req$QUERY_STRING, 0, stringr::str_locate(req$QUERY_STRING, "offset=")[[2]]),
+        max(0, (as.numeric(offset) - as.numeric(limit))),
+        "&limit=", 
+        limit
+      )
+    }
+    
+    return(result)
+  }
+}
+
+#################################################################################################
+
+#' Download the input specified by the id
+#' @param id Input id (character)
+#' @return Input file specified by user
+#' @author Tezan Sahu
+#* @serializer contentType list(type="application/octet-stream")
+#* @get /<input_id>
+downloadInput <- function(input_id, req, res){
+  dbcon <- PEcAn.DB::betyConnect()
+  db_hostid <- PEcAn.DB::dbHostInfo(dbcon)$hostid
+  
+  # This is just for temporary testing due to the existing issue in dbHostInfo()
+  db_hostid <- ifelse(db_hostid == 99, 99000000001, db_hostid)
+  
+  input <- tbl(dbcon, "dbfiles") %>%
+    select(file_name, file_path, container_id, machine_id, container_type) %>%
+    filter(machine_id == !!db_hostid) %>%
+    filter(container_type == "Input") %>%
+    filter(container_id == !!input_id) %>%
+    collect()
+  
+  PEcAn.DB::db.close(dbcon)
+  
+  if (nrow(input) == 0) {
+    res$status <- 404
+    return()
+  }
+  else{
+    # Generate the full file path using the file_path & file_name
+    filepath <- paste0(input$file_path, "/", input$file_name)
+    
+    if(! file.exists(filepath)){
+      res$status <- 404
```

Comment:
Umm...that's a bit inconvenient I guess because the content-type is specified to be `application/octet-stream` as a decorator over the function, and changing that here may not be the easiest thing to do as of now. I think we can keep this on a hold for now & meanwhile I will see if something could be done.

Reply:
Ok. Not the end of the world. Just didn't know if there was a way to send back a message in addition to the raw status code. Definitely not worth holding up this PR for.